ext/bg/js/clipboard-monitor.js

```diff
@@ -55,7 +52,7 @@ class ClipboardMonitor {
             ) {
                 this.previousText = text;
                 if (jpIsStringPartiallyJapanese(text)) {
-                    this.onClipboardText(text);
+                    this.trigger('change', {text});
```

Comment:
I think this would be nicer to use if it was changed to
```javascript
this.trigger('change', [text]);
```

Then you could use
```javascript
this.clipboardMonitor.on('change', (args) => this.onExternalSearchUpdate(...args));
```

and keep
```javascript
onExternalSearchUpdate(text) {
```
and
```javascript
_onClipboardText(text) {
```
as is.

https://github.com/FooSoft/yomichan/issues/359#issuecomment-592934361

Another consistent option would be to go with object destructuring everywhere.

Reply:
My motivation is that then you'll only have to know the arguments at the place where the function is called and where it ends up, not in the middle. It reduces the amount of work and probability of errors when doing changes. I got bitten by it again while doing changes here:

https://github.com/FooSoft/yomichan/pull/389/files#diff-7f1c7cabf21b3d982801b7f56c4f984e

Reply:
The reason I did `{text}` is to mimic an event object, similar to DOM events.
```js
document.body.onclick = (e) => {
  if (e.altKey) { /* do something */ }
};
```

In this case:
```js
something.on('change', (e) => {
  use(e.text);
}
```

Technically I could just pass `text` as the argument without structuring, but IMO using an object is more event-like. Another potential issue with passing directly doesn't apply in this case, but consider if there were more arguments:

```js
// for some event dispatcher "something" and variables text, value1, value2:
something.trigger('change', text, value1, value2);
// ...
something.on('change', (text/*, value1, value2*/) => {
  use(text); // text used; value1, value2 unused and can be omitted
});
something.on('change', (text, value1, value2) => {
  use(value2); // value2 used; text, value1 unused but cannot be omitted
});

// Alternatives which only use what they need without having to declare anything extra
something.trigger('change', {text, value1, value2});
// ...
something.on('change', ({text}) => {
  use(text);
});
something.on('change', ({value2}) => {
  use(value2);
});
```

Reply:
Looks a lot more familiar when there's the `e` :smile: In that light, maybe the approach in this PR is the best way to go, because it addresses the problem with repeating arguments but is just a bit different from other functions around it that don't take in events.

There's a way to make the first argument of a function optional, but it's pretty confusing: https://developer.chrome.com/extensions/runtime#method-sendMessage

I guess that internally checks the type of the first two arguments. I wouldn't want that kind of style.

Reply:
Yeah, in at least one place, the structure of the arguments must be known. Using `.bind` puts this burden on the bound function to ensure it is using the expected argument types. The other places you are mentioning have 3 places which need to know the structure of the arguments: 2 for the handler arrow function and 1 for the target function.

And I agree on early optional function arguments; they are typically ugly and should be avoided since there's no automagic that does it for us.